### PR TITLE
Integrate cuDNN frontend v1 to fused attention

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR:
- [C/PyTorch] integrates cuDNN frontend v1 to fused attention,
- [C/PyTorch] adds support for `padding`, `padding_causal` masks, `post_scale_bias`, `alibi` biases, and MQA/GQA to fused attention,
- [C/PyTorch] widens support for `_qkvpacked` APIs to `h3d/3hd` from `qkv_interleaved`, and support for `_kvpacked` APIs to `hd_h2d/hd_2hd` from `kv_interleaved`,
- [C/PyTorch] removes all usage of `qkv_interleaved`, `kv_interleaved`, `not_interleaved` enums,
- [PyTorch] fixes `FlashAttention` module for `padding_causal` mask,
- [PyTorch] fixes fwd output shape for `thd` format in `FlashAttention`,
- [PyTorch] adds `alibi` support for `UnfusedDPA`,
- [C] changes the backend selection from prioritizing `max512` to `arbitrary_seqlen` backend,
- [PyTorch] changes the backend selection from prioritizing `FlashAttention` to `FusedAttention` `arbitrary_seqlen` backend on sm90,
- [C] disables `padding_causal` mask for cross attention for `max512` backend due to bugs,
- [Pytorch] makes stats_lse contiguous for future context parallel implementations,
- [Pytorch] completely rewrites unit tests in `test_fused_attn.py` for better coverage and efficiency.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -568,8 +570,8 @@ def setup_paddle_extension() -> setuptools.Extension:
     else:
         if version >= (11, 2):
             nvcc_flags.extend(["--threads", "4"])
-        if version >= (11, 0):
-            nvcc_flags.extend(["-gencode", "arch=compute_80,code=sm_80"])
+        #if version >= (11, 0):
+        #    nvcc_flags.extend(["-gencode", "arch=compute_80,code=sm_80"])
```
Comment (ptrendx) [inline setup.py:574]: Why these changes?

Comment (cyanguwa) [reply setup.py:574]: That was only for easy compilation on h100. Will remove them in my final cleanup.

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 2]
```diff
@@ -80,7 +75,8 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
         NVTE_Bias_Type bias_type,
         NVTE_Mask_Type attn_mask_type,
         float dropout, size_t max_seqlen_q,
-        size_t max_seqlen_kv, size_t head_dim) {
+        size_t max_seqlen_kv, size_t head_dim,
+        size_t num_attn_heads, size_t num_gqa_groups) {
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn.cpp:79]: I was wondering if we can keep the order of these arguments aligned for all fused attention API
I noticed there are a few different orders:

1. get_fused_attn_backend: s, d, h, hg
2. max512, arbitrary_qkvpack: b, s, h, d
3. transformer_engine::fused_attn::arbitrary: b, h, hg, s, d
4. arbitrary, arbitrar_kvpack: b, s, h, hg, d

I preferred to use `b, s, h, hg, d` as the order of these arguments and skip one of them if it is not necessary. i.e. `s, h, hg, d` for `get_fused_attn_backend`; `b, s, h, d` for max512

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h [hunk 3]
```diff
@@ -14,53 +14,111 @@
...
+                float scaling_factor, float dropout_probability, NVTE_QKV_Layout layout,
+                NVTE_Bias_Type bias_type, NVTE_Mask_Type mask_type,
+                void* devPtrQ, void* devPtrKTranspose, void* devPtrVTranspose,
+                void* devPtrO, void* devPtrSoftmaxStats,
+                void* devPtrdQ, void* devPtrdK, void* devPtrdV, void* devPtrdO,
+                void* devPtrDropoutSeed, void* devPtrDropoutOffset,
+                void* devPtrCuSeqlensQ, void* devPtrCuSeqlensKV,
+                cudnn_frontend::DataType_t tensorType, void *workspace, size_t *workspace_size,
+                cudaStream_t stream, cudnnHandle_t handle, bool* check_support);
+}  // namespace fused_attn
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h:49]: Why do we need to expose implementation to the header? I supposed the header should only contain the public API, doesn't the implementation only be used in `fused_attn_f16_arbitrary_seqlen.cu`?

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 4]
```diff
@@ -46,1446 +45,608 @@
...
+        || (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
+    bool is_padding = ((mask_type == NVTE_Mask_Type::NVTE_PADDING_MASK)
+        || (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
+    bool is_dropout = (dropout_probability != 0.0f);
+
+    // provide a mechanism for support checking for upcoming FE APIs
+    bool tmp_check = *check_support;
+    if (*check_support && !tmp_check) {
+        *check_support = tmp_check;
+        return;
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:74]: Could you provide more insight on what is this used to? I can't figure out in which situation will go into the branch of line 72.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 5]
```diff
@@ -46,1446 +45,608 @@
...
+            int64_t v_stride[4];
+            generateMatrixStrides(b, h, s_q, s_kv, d, q_stride,
+                    layout, NVTE_QKV_Matrix::NVTE_Q_Matrix);
+            generateMatrixStrides(b, hg, s_q, s_kv, d, k_stride,
+                    layout, NVTE_QKV_Matrix::NVTE_K_Matrix);
+            generateMatrixStrides(b, hg, s_q, s_kv, d, v_stride,
+                    layout, NVTE_QKV_Matrix::NVTE_V_Matrix);
+            std::vector<int64_t> q_strides(q_stride, q_stride + 4);
+            std::vector<int64_t> k_strides(k_stride, k_stride + 4);
+            std::vector<int64_t> v_strides(v_stride, v_stride + 4);
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:134]:
```suggestion
            std::vector<int64_t> q_strides(4);
            std::vector<int64_t> k_strides(4);
            std::vector<int64_t> v_strides(4);
            generateMatrixStrides(b, h, s_q, s_kv, d, q_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_Q_Matrix);
            generateMatrixStrides(b, hg, s_q, s_kv, d, k_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_K_Matrix);
            generateMatrixStrides(b, hg, s_q, s_kv, d, v_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_V_Matrix);

```

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 6]
```diff
@@ -46,1446 +45,608 @@
...
+            generateMatrixStrides(b, hg, s_q, s_kv, d, k_stride,
+                    layout, NVTE_QKV_Matrix::NVTE_K_Matrix);
+            generateMatrixStrides(b, hg, s_q, s_kv, d, v_stride,
+                    layout, NVTE_QKV_Matrix::NVTE_V_Matrix);
+            generateMatrixStrides(b, h, s_q, s_kv, d, o_stride,
+                    layout, NVTE_QKV_Matrix::NVTE_O_Matrix);
+            std::vector<int64_t> q_strides(q_stride, q_stride + 4);
+            std::vector<int64_t> k_strides(k_stride, k_stride + 4);
+            std::vector<int64_t> v_strides(v_stride, v_stride + 4);
+            std::vector<int64_t> o_strides(o_stride, o_stride + 4);
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:415]:
```suggestion
            std::vector<int64_t> q_strides(4);
            std::vector<int64_t> k_strides(4);
            std::vector<int64_t> v_strides(4);
            std::vector<int64_t> o_strides(4);
            generateMatrixStrides(b, h, s_q, s_kv, d, q_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_Q_Matrix);
            generateMatrixStrides(b, hg, s_q, s_kv, d, k_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_K_Matrix);
            generateMatrixStrides(b, hg, s_q, s_kv, d, v_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_V_Matrix);
            generateMatrixStrides(b, h, s_q, s_kv, d, o_strides.data(),
                    layout, NVTE_QKV_Matrix::NVTE_O_Matrix);
```

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 5 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:134]: Done

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 7]
```diff
@@ -91,91 +87,85 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
...
-    if (
-#if (CUDNN_VERSION >= 8903)
-        (sm_arch_ >= 80)
-#else
-        (sm_arch_ == 80 || sm_arch_ == 90)
-#endif
-            && (max_seqlen_q == max_seqlen_kv)
-            && ((head_dim == 64) || (head_dim == 128))
+    if ((cudnn_runtime_version >= 8903 && sm_arch_ >= 80)
+            && (cudnn_runtime_version < 8903 && sm_arch_ == 80 || sm_arch_ == 90)
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp:123]: Thanks for catching that.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 4 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:74]: I was thinking of replacing the content of `nvte_get_fused_attn_backend` with `graph.check_support()` from each backend, so we don't need to maintain the condition checks in `nvte_get_fused_attn_backend` manually. The `check_support` flag can be used as a way to say 'check support please' and then also return the check result. Not the cleanest design. Will have to think about this. Also, v1/prerelease 4's `check_support()` API doesn't do all the checks we currently do yet, so I might just remove this feature for now.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h (hunk 3 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h:49]: This is for the `check_support` feature mentioned below.

File: transformer_engine/common/fused_attn/fused_attn.cpp (hunk 2 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp:79]: I will have to clean this up a bit.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 8]
```diff
@@ -46,1446 +45,600 @@
...
+                cudaStream_t stream, cudnnHandle_t handle, bool* check_support) {
+    NVTE_CHECK_CUDNN(cudnnSetStream(handle, stream));
+
+    bool is_bias = (bias_type == NVTE_Bias_Type::NVTE_POST_SCALE_BIAS);
+    bool is_alibi = (bias_type == NVTE_Bias_Type::NVTE_ALIBI);
+    bool is_causal = ((mask_type == NVTE_Mask_Type::NVTE_CAUSAL_MASK)
+        || (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
+    bool is_padding = ((mask_type == NVTE_Mask_Type::NVTE_PADDING_MASK)
+        || (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
+    bool is_dropout = (dropout_probability != 0.0f);
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:68]: I will do the same on the arbitrary_seqlen backend side.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 9]
```diff
@@ -46,1446 +45,608 @@
...
+        auto plan_workspace_size = mha_graph->get_workspace_size();
 
         // Exit to request upper level API to allocate memory if needed
-        size_t softmaxSum_workspace_size = b * h * s_q * sizeof(float);
-        size_t dqAccum_workspace_size = use_workspace_opt ? 0 : b * s_q * h * d * sizeof(float);
-        size_t actual_seqlen_workspace_size = 2 * b * sizeof(int32_t);
         if (workspace == nullptr) {
-            *workspace_size = plan_workspace_size + softmaxSum_workspace_size
-                              + dqAccum_workspace_size + actual_seqlen_workspace_size;
+            *workspace_size = plan_workspace_size;
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:570]:
```suggestion
            size_t actual_seqlen_workspace_size = 2 * b * sizeof(int32_t);
            *workspace_size = plan_workspace_size + actual_seqlen_workspace_size;
```

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 8 above)
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Done

File: transformer_engine/common/fused_attn/fused_attn.cpp (hunk 2 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp]: Done

**Comment by cyanguwa** [comment]
Pipeline 11107364 with a newer cudnn.

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 10]
```diff
@@ -329,11 +345,17 @@ void nvte_fused_attn_bwd_qkvpacked(
   } else if (fused_attention_backend == NVTE_Fused_Attn_Backend::NVTE_F16_arbitrary_seqlen) {
 #if (CUDNN_VERSION >= 8900)
       Tensor *output_S = reinterpret_cast<Tensor *>(Aux_CTX_Tensors->tensors[0]);
-      const Tensor *input_rng_state = reinterpret_cast<const Tensor*>(Aux_CTX_Tensors->tensors[1]);
+      Tensor *input_Bias, *input_rng_state;
+      if ((bias_type != NVTE_NO_BIAS) && (bias_type != NVTE_ALIBI)) {
+          input_Bias = reinterpret_cast<Tensor*>(Aux_CTX_Tensors->tensors[1]);
+          input_rng_state = reinterpret_cast<Tensor*>(Aux_CTX_Tensors->tensors[2]);
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn.cpp:351]: Could we keep `input_rng_state` in the Aux_CTX_Tensors->tensors[1] to align the other cases?

Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp]: Done. I tried to keep `rng_state` last because our FP8 API is also having it last. Also, the bias tensor is part of the Aux tensor pack because it's flash and bias is required to recompute softmax, so my intention was to keep all the softmax-related tensors together. But if it makes it difficult for Jax to implement, I've swapped them. Thanks.

**Comment by cyanguwa** [comment]
Pipeline 11182007

**Comment by cyanguwa** [comment]
Pipeline 11196472 with newer builds of cuDNN.

File: transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.h [hunk 11]
```diff
@@ -19,7 +19,7 @@
 
 namespace transformer_engine {
 #if (CUDNN_VERSION >= 8901)
-void fused_attn_max_512_fwd_qkvpacked(size_t batch, size_t max_seqlen, size_t num_head,
+void fused_attn_max_512_fwd_qkvpacked(size_t batch, size_t num_head, size_t max_seqlen,
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.h:22]: I'm curious why this PR changes the order of `H` and `S`, I think most of the implementation uses `B, S, H, D` as the layouts. So placing `S` before `H` seems more matching the most of the use cases. Like [NVTE_QKV_Layout](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/include/transformer_engine/fused_attn.h#L33), [jax dot_product_attention](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/_autosummary/flax.linen.dot_product_attention.html), [dao-AILab flash attention](https://github.com/Dao-AILab/flash-attention#how-to-use-flashattention)

File: transformer_engine/common/include/transformer_engine/fused_attn.h [hunk 12]
```diff
@@ -198,8 +149,10 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
                 NVTE_QKV_Layout qkv_layout,
                 NVTE_Bias_Type bias_type,
                 NVTE_Mask_Type attn_mask_type,
-                float dropout, size_t max_seqlen_q,
-                size_t max_seqlen_kv, size_t head_dim);
+                float dropout,
+                size_t max_seqlen_q, size_t max_seqlen_kv,
+                size_t num_attn_heads, size_t num_gqa_groups,
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/include/transformer_engine/fused_attn.h:154]: Fixed.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h [hunk 13]
```diff
@@ -14,53 +14,84 @@
 #include "transformer_engine/fused_attn.h"
 
 #include <cudnn.h>
+#include <cudnn_frontend.h>
+#include <cudnn_frontend_utils.h>
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h:18]: It is better to put these header on the cpp side since there is no cudnn_frontend uses in the header.

File: transformer_engine/common/include/transformer_engine/fused_attn.h [hunk 14]
```diff
@@ -211,14 +164,13 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
  *
  * Support Matrix:
    \verbatim
-   | backend | precision |    qkv layout   |       bias         |                 mask                  | dropout | sequence length | head_dim |
-   | 0       | FP16/BF16 | QKV_INTERLEAVED | NO/POST_SCALE_BIAS | NO_MASK/PADDING/CAUSAL/PADDING_CAUSAL |   Yes   |     <= 512      |    64    |
-   | 1       | FP16/BF16 | QKV_INTERLEAVED |       NO_BIAS      |     PADDING/CAUSAL/PADDING_CAUSAL     |   Yes   |      > 512      |  64, 128 |
-   | 2       | FP8       | QKV_INTERLEAVED |      NO_BIAS       |            PADDING_MASK               |   Yes   |     <= 512      |    64    |
+   | backend | precision | qkv layout |       bias         |                  mask                 | dropout | sequence length | head_dim |
+   | 0       | FP16/BF16 | 3HD or H3D | NO/POST_SCALE_BIAS | NO/PADDING/CAUSAL/PADDING_CAUSAL_MASK |   Yes   |     <= 512      |    64    |
+   | 1       | FP16/BF16 | 3HD or H3D |      NO_BIAS       |          NO/PADDING/CAUSAL_MASK       |   Yes   |      > 512      |   <= 128 |
```
Comment (zlsh80826) [inline transformer_engine/common/include/transformer_engine/fused_attn.h:169]: Doesn't the arbitrary support all mask? The seqlen may be adjust to unlimited or `%32`. Bias is also supported, right?

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 15]
```diff
@@ -91,91 +88,99 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
...
-#else
-        (sm_arch_ == 80 || sm_arch_ == 90)
-#endif
-            && (max_seqlen_q == max_seqlen_kv)
-            && ((head_dim == 64) || (head_dim == 128))
-            && (bias_type == NVTE_Bias_Type::NVTE_NO_BIAS)
+    if (((cudnn_runtime_version >= 8903 && sm_arch_ >= 80)
+                || (cudnn_runtime_version < 8903 && (sm_arch_ == 80 || sm_arch_ == 90)))
+            && (max_seqlen_q % 64 == 0)
+            && (max_seqlen_kv % 64 == 0)
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn.cpp:129]: I remember the seqlen condition is `%32 == 0` after 8.9.3.

Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp:129]: It's supposed to, but there's a bug, so we're still pending on cuDNN fixing that bug.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h (hunk 13 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h:18]: Done.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 16]
```diff
@@ -46,1476 +45,613 @@
...
+                dropout_seed = mha_graph->tensor(fe::graph::Tensor_attributes()
+                                .set_name("Seed")
+                                .set_dim({1, 1, 1, 1})
+                                .set_stride({1, 1, 1, 1})
+                                .set_data_type(fe::DataType_t::INT32));
+                dropout_offset = mha_graph->tensor(fe::graph::Tensor_attributes()
+                                .set_name("Offset")
+                                .set_dim({1, 1, 1, 1})
+                                .set_stride({1, 1, 1, 1})
+                                .set_data_type(fe::DataType_t::INT32));
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Changed back to int64. Thanks!

File: transformer_engine/common/include/transformer_engine/fused_attn.h (hunk 14 above)
Comment (cyanguwa) [reply transformer_engine/common/include/transformer_engine/fused_attn.h]: Fixed some of the columns and prettified it a little bit. Thanks!

**Comment by cyanguwa** [comment]
Pipeline 11267171

**Review by zlsh80826 — APPROVED** [review]
LGTM! Thanks for making this awesome upgrade!

**Comment by cyanguwa** [comment]
Pipeline 11267171 is green. The last 5 commits are mostly to fix L1 tests, with [95d0820](https://github.com/NVIDIA/TransformerEngine/pull/497/commits/95d082022fc9a5428d9bddd551516cddf5fda9fb) an exception. It reduces the amount of prints when training.
